app/pages/learn/01_tutorial/04_mastering-the-api/02_modern_io/index.md

```diff
@@ -0,0 +1,252 @@
+```
+
+Then read the data into a byte array and optionally turn them into a string:
+
+```
+byte[] bytes = in.readAllBytes();
+String result = new String(bytes);
+```
+
+Or transfer the data to an output stream:
+
+```
+OutputStream out = Files.newOutputStream(path);
+in.transferTo(out);
+```
+
+Nowadays, there is no need to read or write bytes, or chunks of bytes, in a loop.
+
+But do you really need an input stream? Many APIs give you the option to read from a file or URL. 
+
+Your favorite JSON library is likely to have methods for reading from a file or URL. For example, with [Jackson jr](https://github.com/FasterXML/jackson-jr):
+
+```
+URL url = new URI("https://dog.ceo/api/breeds/image/random").toURL();
+Map<String, Object> result = JSON.std.mapFrom(url);
+```
+
+Here is how to read the dog image from the preceding call:
+
+```
+url = new URI(result.get("message").toString()).toURL();
+BufferedImage img = javax.imageio.ImageIO.read(url)
+```
+
+This is better than passing an input stream to the `read` method, because the library can use additional information from the URL to determine the image type.
+
+## The Files API
+
+The `java.nio.file.Files` class provides a comprehensive set of file operations, such as creating, copying, moving, and deleting fies and directories. The [File System Basics](https://dev.java/learn/java-io/file-system/) tutorial provides a thorough description. In this section, I highlight a few common tasks.
+
+### Traversing Entries in Directories and Subdirectories
+
+For most situations you can use one of two methods. The `Files.list` method visits all entries (files, subdirectories, symbolic links) of a directory.
+
+```
+try (Stream<Path> entries = Files.list(pathToDirectory)) 
+{
+   . . .
+}
+```
+
+Use a `try`-with-resources statement to ensure that the stream object, which keeps track of the iteration, will be closed.
+
+If you also want to visit the entries of descendant directories, instead use the method
+
+```
+Stream<Path> entries = Files.walk(pathToDirectory);
+```
+
+Then simply use stream methods to home in on the entries that you are interested in, and to collect the results:
+
+```
+try (Stream<Path> entries = Files.walk(pathToDirectory)) {
+   List<Path> htmlFiles = entries.filter(p -> p.toString().endsWith("html")).toList();
+   . . .
+}
+```
+
+Here are the other methods for traversing directory entries:
+
+* An overloaded version of `Files.walk` lets you limit the depth of the traversed tree.
+* Two `Files.walkFileTree` methods provide more control over the iteration process, by notifying a `FileVisitor` when a directory is visited for the first and last time. This can be occasionally useful, in particularly for emptying and deleting a tree of directories. See the tutorial [=https://dev.java/learn/java-io/file-system/walking-tree/ Walking the File Tree](null) for details. Unless you need this control, use the simpler `Files.walk` method.
```

Comment:
Is the `[=https://dev.java/learn/java-io/file-system/walking-tree/ Walking the File Tree](null)` intended here?

Reply:
Fixed